Sources/CSRF.swift

```diff
@@ -57,139 +50,59 @@ public struct CSRF: Middleware {
+            let random = CryptoRandom()
+            let secretData = try random.generateData(count: 16)
+            let secret = secretData.hexEncodedString()
+            session["CSRFSecret"] = secret
             return secret
         }
         
         return secret
     }
     
-    private func defaultTokenRetrieval(from request: Request) throws -> String {
-        if let token = request.parameters["_csrf"]?.string {
-            return token
-        }
+    private static func defaultTokenRetrieval(from request: Request) throws -> Future<String> {
         
+        let promise = request.eventLoop.newPromise(String.self)
         let csrfKeys: Set<String> = ["_csrf", "csrf-token", "xsrf-token", "x-csrf-token", "x-xsrf-token", "x-csrftoken"]
-        let requestHeaderKeys = Set(request.headers.keys.map { $0.key })
+        let requestHeaderKeys = Set(request.http.headers.map { $0.name })
         
         let intersection = csrfKeys.intersection(requestHeaderKeys)
         
-        guard let match = intersection.first else {
-            throw Abort(.forbidden,
-                        metadata: nil,
-                        reason: "No CSRF token provided.",
-                        identifier: nil,
-                        possibleCauses: ["Perhaps you forgot to create a token for the session.",
-                                         "You could be using a different header key for the CSRF token than is covered by the default token retrieval."],
-                        suggestedFixes: ["Make sure to create and add a token to your request header. \nSee `CSRF.createToken(from:) for documentation on usage.",
-                                         "Look at `CSRF.defaultTokenRetrieval(from:)` to see what keys are looked for by default."],
-                        documentationLinks: nil,
-                        stackOverflowQuestions: nil,
-                        gitHubIssues: nil)
+        if let matchingKey = intersection.first, let token = request.http.headers[matchingKey].first {
+            promise.succeed(result: token)
+            return promise.futureResult
         }
         
-        let matchingKey = HeaderKey(match)
-        guard let token = request.headers[matchingKey]?.string else {
-            throw Abort(.forbidden,
-                        metadata: nil,
-                        reason: "Failed to find token for key: \(matchingKey).",
-                        identifier: nil,
-                        possibleCauses: nil,
-                        suggestedFixes: nil,
-                        documentationLinks: nil,
-                        stackOverflowQuestions: nil,
-                        gitHubIssues: nil)
-        }
-        return token
-    }
-}
-
-extension CSRF {
-    
-    /// An `OptionSet` representing the varous HTTP methods to ignore.
-    public struct HTTPMethod: OptionSet {
-        public let rawValue: Int
+        request.content.get(at: "_csrf")
```

Comment:
Is there ever a reason `defaultTokenRetrieval` would return a return? You can change this to a `sync` get to remove all the async complexity from this method

Reply:
I can remove async complexity from this method, but I think that TokenRetrievalHandler should be async. Therefore, how about changing this method to sync and wrapping ```request.future```?

Reply:
It depends what `TokenRetrievalHandler` is going to do - if it only ever returns a token from a request I would argue that's probably never going to be asynchronous - it's not streaming or waiting from a response from the DB etc

Reply:
If getting a token only from a request, I think there are no issues with sync processing. I'll fix.

Reply:
Sorry, since `syncGet` uses DataDecoder, a token can not be retrieved when ContentType is `multipart/form-data`. Therefore, I think that TokenRetrievalHandler must be returing async type. What about you?

Reply:
Ah that’s a good point. Yeah it’ll have to be asynchronous then